src/Type/Equality.purs

```diff
@@ -14,10 +16,26 @@ module Type.Equality
 -- | Note: any instance will necessarily overlap with
 -- | `refl` below, so instances of this class should
 -- | not be defined in libraries.
+class TypeEquals :: forall k. k -> k -> Constraint
 class TypeEquals a b | a -> b, b -> a where
-  to :: a -> b
-  from :: b -> a
+  leibniz :: forall p. p a -> p b
 
 instance refl :: TypeEquals a a where
-  to a = a
-  from a = a
+  leibniz a = a
+
+newtype Op :: forall k. k -> k -> Type
+newtype Op a b = Op (forall p. p b -> p a)
+
+leibnizOp :: forall p a b. TypeEquals a b => p b -> p a
+leibnizOp = case leibniz (Op (\pa -> pa) :: Op a a) of
+  (Op f :: Op a b) -> f
+
+newtype To a = To a
+
+to :: forall a b. TypeEquals a b => a -> b
+to a = case leibniz (To a) of
+  To b -> b
+
+from :: forall a b. TypeEquals a b => b -> a
+from b = case leibnizOp (To b) of
+  To a -> a
```

Comment:
Is `leibnizOp` really necessary? Given `Data.Op.Op` we could define `from` as `case leibniz (Op \b -> b) of Op f -> f`.

If we don’t want any dependency `newtype Op a b = Op (b -> a)` is also perhaps a bit simpler than `newtype Op a b = Op (forall p. p b -> p a)` and could rename it to `Symm` or something more semantic for this use case (the reversal of a leibniz equality).

Reply:
I think having the symmetric/opposite form of Leibniz in this place is good, because sometimes you’ll need both directions (just for easier access) and you don’t want to have to add another constraint. It’s just one newtype that only needs to be used internally.

Reply:
I think I would prefer not to export leibnizOp at first; if people want to flip it around they can use the leibniz library, right? If we’re saying this is only the basic stuff and people should use the leibniz library for more involved usage, then to me that seems consistent with not exporting leibnizOp.